src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    /**
+     * Asynchronously closes each writer in the pool. Each writer calls {@code .close()} in a CompletableFuture, which
+     * allows the writers to wait for any ongoing writes, drain any remaining elements from the queue, and then close
+     * the inner writer. All writers will immediately cease to accept new items to write and any call to {@code .write()}
+     * after calling this method will throw an exception. This method waits for all {@code CompletableFuture}s to
+     * complete and then shuts down the executor.
+     * <p>
+     * This method blocks and till all writers have closed and pool has shut down.
+     *
+     * @throws IOException if any writer raises an exception
+     */
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        List<CompletableFuture<Void>> closers = new ArrayList<>();
+        for (final PooledWriter<?> writer : this.writers) {
+            closers.add(writer.asyncClose());
+        }
+        // Convert List to array and wait on all futures to complete
+        CompletableFuture.allOf(closers.toArray(new CompletableFuture[0])).join();
+        this.executor.shutdown();
+    }
+
+    /**
+     * Exchange a class implementing {@link Writer} for a {@link PooledWriter}.
+     *
+     * @param writer         a class implementing {@link Writer}
+     * @param queue          a queue to use for this writer, bounded or unbounded.
+     * @param writeThreshold the minimum number of items needed before scheduling a thread for writing.
+     * @param <A>            the type of the items the {@link Writer} can write.
+     * @return a {@link PooledWriter}
+     */
+    public <A> PooledWriter<A> pool(final Writer<A> writer, final BlockingQueue<A> queue, final int writeThreshold) {
+        PooledWriter<A> pooledWriter = new PooledWriter<>(writer, queue, writeThreshold);
+        this.writers.add(pooledWriter);
+        return pooledWriter;
+    }
+
+
+    /**
+     * Any class that implements {@link Writer} can be exchanged for a {@code PooledWriter}. The PooledWriter provides
+     * the same API as {@link Writer}, but will manage buffering of writes and sending to the {@link AsyncWriterPool} it
+     * is attached to.
+     * <p>
+     * Note, {@code PooledWriter} is not intrinsically threadsafe, and any writing / closing done across multiple
+     * threads should be synchronized.
+     *
+     * @param <A> the type of the item the writer can write
+     */
+    public class PooledWriter<A> implements Writer<A> {
+        private final AtomicBoolean isClosed = new AtomicBoolean(false);
+        private final BlockingQueue<A> queue;
+        private final Writer<A> writer;
+        private final int writeThreshold;
+
+        // Holds the Future of the last task submitted to the AsyncWriterPools until it is checked, then it is null again.
+        private final AtomicReference<Future<Void>> currentTask = new AtomicReference<>(null);
```

Comment:
I'm 99% sure that this doesn't need to be wrapped in an AtomicReference.  Since `PooledWriter` isn't supposed to be thread safe, users of it should either a) be invoking methods from a single thread or b) syncrhonizing externally.  And I _think_ the `currentTask` is only updated in methods invoked by the user, not from within the executor threads.

Reply:
`currentTask` is updated by calls to `drain` and calls to `checkAndRethrow`. The only time those methods can be called from another thread is during the `nonBlockingClose`. I believe that with the newest changes to `nonBlockingClose` we remain in a scenario where only one thread can access `currentTask`.